### PR TITLE
JSAB2-101 Setup entire backend

### PR DESCRIPTION
**Pull request by kyya**
+ setup mongo database
+ setup some models
+ deploy the database to mongo atlas

### CODE REVIEW

**Review by HanqingZ — APPROVED** [review]

**Review by GuJialu — APPROVED** [review]

**Review by HONGTAO-T — APPROVED** [review]

File: docs/README.md [hunk 1]
```diff
@@ -0,0 +1,204 @@
...
+POST /kingdom/buildings/:buildingType
+```
+
+| param        | Description                                                |
+| ------------ | ---------------------------------------------------------- |
+| buildingType | The type of building have been created<br />{ Farm, Mine } |
+
+#### Get a building with id
+
+description
```
Comment (evasimonyi) [inline docs/README.md]: please write a short description (like line 97, something like that is enough) or delete this line, but don't leave `description` here

File: docs/README.md [hunk 2]
```diff
@@ -0,0 +1,204 @@
...
+POST /kingdom/buildings/:buildingId
+```
+
+| param      | Description                               |
+| ---------- | ----------------------------------------- |
+| buildingId | The ID of selected building to be upgrade |
+
+#### Upgrade a building with id
+
+description
```
Comment (evasimonyi) [inline docs/README.md]: please write a short description (like line 97, something like that is enough) or delete this line, but don't leave `description` here

File: docs/README.md [hunk 3]
```diff
@@ -0,0 +1,204 @@
...
+
+| param | Description                                   |
+| ----- | --------------------------------------------- |
+| level | The number of level to create a unit of troop |
+
+
+
+#### Upgrade a unit of troop
+
+description
```
Comment (evasimonyi) [inline docs/README.md]: please write a short description (like line 97, something like that is enough) or delete this line, but don't leave `description` here

File: docs/README.md [hunk 4]
```diff
@@ -0,0 +1,204 @@
...
+
+| param | Description                                    |
+| ----- | ---------------------------------------------- |
+| level | The number of level to upgrade a unit of troop |
+
+
+
+#### Start a battle between troops
+
+description
```
Comment (evasimonyi) [inline docs/README.md]: please write a short description (like line 97, something like that is enough) or delete this line, but don't leave `description` here

File: docs/README.md [hunk 5]
```diff
@@ -0,0 +1,204 @@
...
+PUT /kingdom/troops/upgrade?level=1
+```
+
+| param | Description                                    |
+| ----- | ---------------------------------------------- |
+| level | The number of level to upgrade a unit of troop |
+
+
+
+#### Start a battle between troops
```
Comment (evasimonyi) [inline docs/README.md]:
```suggestion
#### Start a battle between kingdoms
```

it is fought with troops, but the fight is between kingdoms, right?

**Review by GuJialu — APPROVED** [review]

File: backend/App.js [hunk 6]
```diff
@@ -1,12 +1,13 @@
+require('dotenv').config();
 const express = require('express');
-
-const app = express();
-const PORT = 4000;
 const {
-  troops, buildings, resources, notification,
+  users, troops, buildings, resources, notification, progresses,
```
Comment (evasimonyi) [inline backend/App.js]: it is the progress of the construction of a building/troop, or upgrading them, or it is a general progress function, and we'll use this for notifications/progress bars/construction logs? just a question, thanks

File: backend/routers/troops.js [hunk 7]
```diff
@@ -1,61 +1,37 @@
...
+      countByLevel,
+    });
+    res.send(result);
+  } catch (error) {
+    next(error);
+  }
 }
 
 router.get('/', getTroops);
+router.post('/', createTroop);
```
Comment (evasimonyi) [inline backend/routers/troops.js]: please apply `auth` middleware

File: backend/routers/troops.js [hunk 8]
```diff
@@ -1,61 +1,37 @@
...
+      owner,
+      countByLevel,
+    });
+    res.send(result);
+  } catch (error) {
+    next(error);
+  }
 }
 
 router.get('/', getTroops);
```
Comment (evasimonyi) [inline backend/routers/troops.js]: please apply `auth` middleware

File: backend/routers/buildings.js [hunk 9]
```diff
@@ -1,97 +1,100 @@
...
+  } catch (error) {
+    next(error);
   }
-  return res.sendStatus(404);
 }
 
-router.get('/', getBuildings);
+router.get('/', auth, getBuildings);
+router.post('/:buildingType', auth, createBuilding);
 router.get('/:buildingId', getBuildingById);
```
Comment (evasimonyi) [inline backend/routers/buildings.js]: please apply `auth` middleware on every endpoint besides login and register

File: backend/models/progress.js [hunk 10]
```diff
@@ -0,0 +1,31 @@
...
+const progressSchema = new Schema({
+  owner: { type: Schema.Types.ObjectId, ref: 'User' },
+  category: String,
+  type: String,
+  toLevel: Number,
+  startAt: Number,
+  finishAt: Number,
+}, schemaOptions);
+
+progressSchema.virtual('isUpgrade').get(function () {
```
Comment (evasimonyi) [inline backend/models/progress.js]:
```suggestion
progressSchema.virtual('isUpgraded').get(function () {
```

File: backend/routers/users.js [hunk 11]
```diff
@@ -0,0 +1,142 @@
...
+    }
+    const passwordHash = await bcrypt.hash(password, 8);
+    const result = await UserModel.create({
+      email,
+      username,
+      password: passwordHash,
+      kingdomName: kingdomName || `${username}'s kingdom`,
+      kingdomList: [],
+    });
+    res.status(200).send(result);
```
Comment (evasimonyi) [inline backend/routers/users.js]: NEVER send back the password, not even a hashed one!

File: backend/App.js [hunk 12]
```diff
@@ -1,12 +1,13 @@
...
 } = require('./routers');
 
+const app = express();
+
 app.use(express.json());
+app.use(express.urlencoded({ extended: true }));
 
 app.get('/', (req, res) => {
```
Comment (evasimonyi) [inline backend/App.js:12]: please delete the `/` endpoint, it's unnecessary now

File: backend/routers/buildings.js [hunk 13]
```diff
@@ -1,97 +1,100 @@
...
-    }, {
-      id: 4,
-      type: 'Mine',
-      level: 1,
-      hp: 1,
-      started_at: 12345789,
-      finished_at: 12399999,
-    },
-  ],
+const typeToRules = {
```
Comment (evasimonyi) [inline backend/routers/buildings.js]: `rulesOfBuildings` or `buildingRules` or something like that maybe?

File: backend/routers/buildings.js [hunk 14]
```diff
@@ -1,97 +1,100 @@
...
+    const moneyRequired = parseInt(typeToRules[buildingType].constructionCost, 10);
+    const resourceOfUser = await ResourceModel.find({ owner });
+    const [moneyInHand] = resourceOfUser
+      .filter(({ type }) => (type === 'gold'))
+      .map(({ amount }) => amount);
+    if (moneyRequired > moneyInHand) {
+      throw createError(400, 'You don\'t have enough money');
+    }
+
+    const minusGold = await ResourceModel.findOneAndUpdate(
```
Comment (evasimonyi) [inline backend/routers/buildings.js]: this is not descriptive enough. the value of this variable is the amount you want to subtract or the amount the user is left AFTER we subtract the spent amount? it seems that it's the latter, so I'd rather call it `currentAmountOfGold` or `updatedAmountOfGold` or something similar.

File: backend/routers/buildings.js [hunk 15]
```diff
@@ -1,97 +1,100 @@
...
+  try {
+    if (!buildingType) {
+      throw createError(400, 'Type is required');
+    }
+    if (buildingType !== 'Mine' && buildingType !== 'Farm') {
+      throw createError(400, 'Wrong type');
+    }
+
+    const moneyRequired = parseInt(typeToRules[buildingType].constructionCost, 10);
+    const resourceOfUser = await ResourceModel.find({ owner });
```
Comment (evasimonyi) [inline backend/routers/buildings.js]:
```suggestion
    const resourcesOfUser = await ResourceModel.find({ owner });
```

File: backend/routers/buildings.js [hunk 16]
```diff
@@ -1,97 +1,100 @@
...
+    const minusGold = await ResourceModel.findOneAndUpdate(
+      { owner, type: 'gold' },
+      { $inc: { amount: -{ moneyRequired } } },
+      { new: true, fields: '-_id' },
+    ).exec();
+    const createNewBuilding = await BuildingModel.create({
+      type: buildingType,
+      owner,
+    });
+    res.send({ minusGold, createNewBuilding });
```
Comment (evasimonyi) [inline backend/routers/buildings.js]:
```suggestion
    res.send({ minusGold, newBuilding });
```

if the name contains 'create' in it, it suggests it's a function, but we want to send back a building now, and not a function

File: backend/routers/buildings.js [hunk 17]
```diff
@@ -1,97 +1,100 @@
...
+  try {
+    const building = await BuildingModel.findById(id);
+    if (!building) {
+      throw createError(404, 'This building doesn\'t exist.');
     }
+    res.send({
+      building,
+      rules: {
+        buildingRules: typeToRules[building.type],
+        troopsRules: foxRule,
```
Comment (evasimonyi) [inline backend/routers/buildings.js:70]: why are you sending back the troops rules too with the building?

File: backend/routers/buildings.js [hunk 18]
```diff
@@ -1,97 +1,100 @@
...
+      },
+    });
+  } catch (error) {
+    next(error);
+  }
+}
+
+async function upgradeBuildingById(req, res, next) {
```
Comment (evasimonyi) [inline backend/routers/buildings.js:78]: you don't check if the Townhall level is above the level that you want to upgrade the given building onto
neither that you have enough money for it
neither you subtract the spent money

please fix it

PS: these are used multiple times (checking if the user has enough money, and subtract it), (constructing new buildings and troops, upgrading buildings and troops), maybe these could be more generic functions?

File: backend/routers/buildings.js [hunk 19]
```diff
@@ -1,97 +1,100 @@
...
+      throw createError(400, 'Wrong type');
+    }
+
+    const moneyRequired = parseInt(typeToRules[buildingType].constructionCost, 10);
+    const resourceOfUser = await ResourceModel.find({ owner });
+    const [moneyInHand] = resourceOfUser
+      .filter(({ type }) => (type === 'gold'))
+      .map(({ amount }) => amount);
+    if (moneyRequired > moneyInHand) {
+      throw createError(400, 'You don\'t have enough money');
```
Comment (evasimonyi) [inline backend/routers/buildings.js:43]: this checking function could be used in multiple cases through the app, I think you should write this outside of the `createBuilding` function

File: backend/routers/buildings.js [hunk 20]
```diff
@@ -1,97 +1,100 @@
...
+    }
+
+    const minusGold = await ResourceModel.findOneAndUpdate(
+      { owner, type: 'gold' },
+      { $inc: { amount: -{ moneyRequired } } },
+      { new: true, fields: '-_id' },
+    ).exec();
+    const createNewBuilding = await BuildingModel.create({
```
Comment (evasimonyi) [inline backend/routers/buildings.js]:
```suggestion
    const newBuilding = await BuildingModel.create({
```

File: backend/routers/resources.js [hunk 21]
```diff
@@ -1,41 +1,24 @@
...
-  const calculatedResources = {
-    resources: mockedResources.resources.map((resource) => ({
-      type: resource.type,
-      amount: getResourceAmount(resource.amount, resource.generation, resource.updatedAt),
-      generation: resource.generation,
-    })),
-  };
-  return res.status(200).send(calculatedResources);
+async function getResources(req, res, next) {
+  const { _id: owner } = req.user;
```
Comment (evasimonyi) [inline backend/routers/resources.js:8]: why do you have this here and in `getBuildings` in `routers/buildings.js` but not in the `getTroops` function in the `routers/troops.js` file?

File: backend/routers/buildings.js [hunk 22]
```diff
@@ -1,97 +1,100 @@
...
+    next(error);
   }
-  return res.sendStatus(404);
 }
 
-router.get('/', getBuildings);
+router.get('/', auth, getBuildings);
+router.post('/:buildingType', auth, createBuilding);
 router.get('/:buildingId', getBuildingById);
+router.post('/:buildingId/upgrade', upgradeBuildingById);
```
Comment (evasimonyi) [inline backend/routers/buildings.js]:
```suggestion
router.patch('/:buildingId', upgradeBuildingById);
```

File: backend/routers/buildings.js [hunk 23]
```diff
@@ -1,97 +1,100 @@
...
+    res.status(200).send(result);
+  } catch (error) {
+    next(error);
   }
-  return res.sendStatus(404);
 }
 
-router.get('/', getBuildings);
+router.get('/', auth, getBuildings);
+router.post('/:buildingType', auth, createBuilding);
```
Comment (evasimonyi) [inline backend/routers/buildings.js]:
```suggestion
router.post('/', auth, createBuilding);
```

Comment (evasimonyi) [reply backend/routers/buildings.js]: REST
[https://en.wikipedia.org/wiki/Representational_state_transfer#Relationship_between_URI_and_HTTP_methods](url)

File: backend/routers/buildings.js (hunk 22 above)
Comment (evasimonyi) [reply backend/routers/buildings.js]: REST
[https://en.wikipedia.org/wiki/Representational_state_transfer#Relationship_between_URI_and_HTTP_methods](url)

File: backend/routers/users.js [hunk 24]
```diff
@@ -0,0 +1,142 @@
...
+    res.status(202).send({ message });
+  } catch (error) {
+    next(error);
+  }
+}
+
+router.get('/:uid?', getUser);
+router.post('/login', login);
+router.post('/register', register);
+router.patch('/setting', auth, resetUserInfo);
```
Comment (evasimonyi) [inline backend/routers/users.js]:
```suggestion
router.patch('/settings', auth, resetUserInfo);
```

File: backend/routers/users.js [hunk 25]
```diff
@@ -0,0 +1,142 @@
...
+
+    user.password = undefined;
+    const token = jwt.sign({ user }, process.env.APP_SECRET || 'testSecret');
+    res.status(200).send({ token });
+  } catch (error) {
+    next(error);
+  }
+}
+
+async function resetUserInfo(req, res, next) {
```
Comment (evasimonyi) [inline backend/routers/users.js]:
```suggestion
async function updateUserInfo(req, res, next) {
```

`update` or `modify` or `change`
but no reset
reseting means we reset their data to the state when they registered, but that's not the case

File: backend/tests/buildings.test.js [hunk 26]
```diff
@@ -1,46 +1,38 @@
 const request = require('supertest');
 const app = require('../App');
 
+const token = [
```
Comment (evasimonyi) [inline backend/tests/buildings.test.js]: just FYI:

you can use `before`, and use the login function here with a mocked (but actually registered) user (like Kyya - password), so it will login and generate a token

[https://mochajs.org/#hooks](url)

**Review by evasimonyi — CHANGES_REQUESTED** [review]

File: backend/routers/troops.js (hunk 7 above)
Comment (GuJialu) [reply backend/routers/troops.js]: This is done in JSAB2-126, which hasn't been merged yet.

File: backend/routers/troops.js (hunk 8 above)
Comment (GuJialu) [reply backend/routers/troops.js]: This is done in JSAB2-126, which hasn't been merged yet.

File: backend/models/progress.js (hunk 10 above)
Comment (GuJialu) [reply backend/models/progress.js]: This field is used to identify whether this progress is used to construct a new unit or upgrade an existing unit. Will it be better to name it 'isUpgrading'?

File: backend/tests/buildings.test.js (hunk 26 above)
Comment (evasimonyi) [reply backend/tests/buildings.test.js]: this is a way too
https://github.com/green-fox-academy/jsa-bloodstone-app/blob/2afea713793fa56decb27b6e2a2a05c96891b576/backend/tests/troops.test.js

File: backend/App.js (hunk 6 above)
Comment (GuJialu) [reply backend/App.js]: This is the progress table stores all the units including buildings and troops that are under construction or upgrading. When the front fetching the units, the back end will first check is there any finished units in the progress table. If so, it will add the finished ones in the buildings/troops table.

File: backend/routers/buildings.js (hunk 17 above)
Comment (HanqingZ) [reply backend/routers/buildings.js:70]: Cause for the frontend of the one building part, we will present all the values of the building and troops from the rules.

File: backend/routers/resources.js (hunk 21 above)
Comment (HanqingZ) [reply backend/routers/resources.js:8]: Jialu already fix it

File: backend/routers/buildings.js (hunk 18 above)
Comment (HanqingZ) [reply backend/routers/buildings.js:78]: the function of purchase with gold is written inside of file model/resource.js

File: backend/routers/buildings.js (hunk 17 above)
Comment (evasimonyi) [reply backend/routers/buildings.js:70]: great, thanks!

**Review by HanqingZ — APPROVED** [review]

File: backend/App.js (hunk 12 above)
Comment (HanqingZ) [reply backend/App.js:12]: we need to test whether our endpoint is right. Just to make sure we can go to this endpoint.

**Review by HONGTAO-T — APPROVED** [review]

File: .circleci/config.yml [hunk 27]
```diff
@@ -36,6 +36,3 @@ jobs:
       - run:
           name: Run Backend ESLint Check
           command: 'cd backend && yarn lint'
-      - run:
-          name: Run Backend Tests
-          command: 'cd backend && yarn test'
```
Comment (evasimonyi) [inline .circleci/config.yml:41]: why did you delete this here, but not the `test` script from the package json? isn't it needed?

**Review by HONGTAO-T — APPROVED** [review]

**Review by GuJialu — APPROVED** [review]

File: backend/App.js (hunk 12 above)
Comment (evasimonyi) [reply backend/App.js:12]: now it's OK but in production there shouldn't be any 'Hello World'-s

File: backend/routers/users.js [hunk 28]
```diff
@@ -0,0 +1,144 @@
...
+    }
+    const passwordHash = await bcrypt.hash(password, 8);
+    const user = await UserModel.create({
+      email,
+      username,
+      password: passwordHash,
+      kingdomName: kingdomName || `${username}'s kingdom`,
+      kingdomList: [],
+    });
+    user.password = undefined;
```
Comment (evasimonyi) [inline backend/routers/users.js]:
```suggestion
    delete user.password;
```

instead of replacing the value to `undefined`, delete it from the user object, so it won't be in the token at all

File: backend/routers/users.js [hunk 29]
```diff
@@ -0,0 +1,144 @@
...
+    const token = jwt.sign({ user }, process.env.APP_SECRET || 'testSecret');
+    res.status(201).send({ token });
```
Comment (evasimonyi) [inline backend/routers/users.js:70]: in our app the user will be redirected to the login page after registration, so here we don't need to send back a token
if you want to redirect him to the logged in main page in the game, we can keep this. please decide. If we keep this, we have to redirect the user to the main page instead of the login page.

File: backend/routers/users.js [hunk 30]
```diff
@@ -0,0 +1,144 @@
...
+    }
+    if (!password) {
+      throw createError(400, 'Password is required.');
+    }
+    const user = await UserModel.findOne({ username });
+    if (!user || !await bcrypt.compare(password, user.password)) {
+      throw createError(400, 'Username or password is incorrect.');
+    }
+
+    user.password = undefined;
```
Comment (evasimonyi) [inline backend/routers/users.js]:
```suggestion
    delete user.password;
```

File: backend/routers/users.js [hunk 31]
```diff
@@ -0,0 +1,144 @@
...
+    const user = await UserModel.findOneAndUpdate(
+      { username: oldUsername },
+      { $set: changedTarget },
+      { new: true, fields: '-_id' },
+    );
+    if (!user) {
+      throw createError(400, 'Can\'t find this username in database');
+    }
+    const message = `${changedValue.join(', ')} are successfully changed!`;
+    res.status(202).send({ message });
```
Comment (evasimonyi) [inline backend/routers/users.js]: for now it's not a problem that you send back this text, but for the future: it is better to send back only the status code in this case, and write the message to be rendered on the frontend, so if we want to change the language of the page, we don't have to do the translation on the backend and on the frontend too, only on the frontend. we shouldn't use these messages on the frontend, we should decide what text to render on the frontend based on the status code.

File: backend/tests/buildings.test.js [hunk 32]
```diff
@@ -2,45 +2,43 @@ const request = require('supertest');
...
-      request(app)
-        .get('/kingdom/buildings/-1')
-        .expect(400, done);
+        .expect(403, done);
     });
   });
 
-  describe('GET /kingdom/buildings/0', () => {
-    it('should return 400 when trying to get a building with id 0', (done) => {
+  describe('GET /kingdom/buildings/5e0a1ceabadb94131493817a', () => {
```
Comment (evasimonyi) [inline backend/tests/buildings.test.js]: just a question: why is this the id of the building, and not 1, 2, etc? it looks like this in the database?

File: backend/tests/buildings.test.js [hunk 33]
```diff
@@ -2,45 +2,43 @@ const request = require('supertest');
...
+
+  before((done) => {
+    request(app)
+      .post('/users/login')
+      .send({ username: 'testuser2', password: '12345678' })
+      .end((err, res) => {
+        token = res.body.token;
+        done();
+      });
+  });
```
Comment (evasimonyi) [inline backend/tests/buildings.test.js]: code duplication: this function could be in a separate `loginUser.js` file (or something similar naming) in the test folder, and be imported in every test file. this way if we delete testuser2 somehow, we can use another test user easily.

**Review by evasimonyi — CHANGES_REQUESTED** [review]

File: .circleci/config.yml (hunk 27 above)
Comment (HanqingZ) [reply .circleci/config.yml:41]: we'll delete the test script

Comment (evasimonyi) [reply .circleci/config.yml:41]: @HanqingZ why delete it here if we put it back on the testing branch as far as I see?

File: backend/routers/users.js (hunk 29 above)
Comment (HanqingZ) [reply backend/routers/users.js:70]: we decide that after registration, the page will directly jump to the menu page. So we'll change the frontend part only

File: backend/tests/buildings.test.js (hunk 32 above)
Comment (HanqingZ) [reply backend/tests/buildings.test.js]: @GuJialu, he said that we only need to use the id given by the mongoose. and the mongoose id looks like this -> 5e0a1ceabadb94131493817a

Comment (GuJialu) [reply backend/tests/buildings.test.js]: Because naturally mongodb uses string id. It is an unnecessary layer to convert the id to numbers. Furthermore, the plugin we found to convert the string id to number id only has 5000 downloads per week.

File: backend/routers/users.js (hunk 31 above)
Comment (HanqingZ) [reply backend/routers/users.js]: I think I'll send this message back to the front as a toast. So that users will know which part already been changed.

Comment (evasimonyi) [reply backend/routers/users.js]: just sayin, that the text that is rendered (in the toast, or in an alert, or in a component) should be written on the frontend side based on the status code (except error messages), so it's ok now if you want to send it back, but in the future it's not the best practice.
also if the user changes only one thing, there should be an `is` instead of the `are` in the message

File: .circleci/config.yml (hunk 27 above)
Comment (HanqingZ) [reply .circleci/config.yml:41]: Sorry, we already delete them, and will update in the test branch

**Review by evasimonyi — APPROVED** [review]
👍
